File: bms/go/ach/moov_io_ach_EntryDetail_TraceNumberField/moov_io_ach_EntryDetail_TraceNumberField.go
package test

import (
	"strings"
	"unicode/utf8"
)

var (
	moov_io_ach_stringZeros map[int]string = moov_io_ach_populateMap(94, "0")
)

const ()

func (ed *moov_io_ach_EntryDetail) TraceNumberField() string {
	return ed.stringField(ed.TraceNumber, 15)
}

func (c *moov_io_ach_converters) stringField(s string, max uint) string {
	ln := uint(utf8.RuneCountInString(s))
	if ln > max {
		return s[:max]
	}

	// Pad with preallocated string
	m := int(max - ln)
	pad, exists := moov_io_ach_stringZeros[m]
	if exists {
		return pad + s
	}
	// slow path
	return strings.Repeat("0", m) + s
}

func moov_io_ach_populateMap(max int, zero string) map[int]string {
	out := make(map[int]string, max)
	for i := 0; i < max; i++ {
		out[i] = strings.Repeat(zero, i)
	}
	return out
}

type moov_io_ach_EntryDetail struct {
	// ID is a client defined string used as a reference to this record.

	// TransactionCode if the receivers account is checking, savings, general ledger (GL) or loan.

	// RDFIIdentification is the RDFI's routing number without the last digit.
	// Receiving Depository Financial Institution

	// CheckDigit the last digit of the RDFI's routing number

	// DFIAccountNumber is the receiver's bank account number you are crediting/debiting.
	// It important to note that this is an alphanumeric field, so its space padded, no zero padded

	// Amount Number of cents you are debiting/crediting this account

	// IdentificationNumber an internal identification (alphanumeric) that
	// you use to uniquely identify this Entry Detail Record

	// IndividualName The name of the receiver, usually the name on the bank account

	// DiscretionaryData allows ODFIs to include codes, of significance only to them,
	// to enable specialized handling of the entry. There will be no
	// standardized interpretation for the value of this field. It can either
	// be a single two-character code, or two distinct one-character codes,
	// according to the needs of the ODFI and/or Originator involved. This
	// field must be returned intact for any returned entry.
	//
	// WEB and TEL batches use the Discretionary Data Field as the Payment Type Code

	// AddendaRecordIndicator indicates the existence of an Addenda Record.
	// A value of "1" indicates that one ore more addenda records follow,
	// and "0" means no such record is present.

	// TraceNumber is assigned by the ODFI or software vendor and used as part of identification.
	//
	// The format of trace numbers is the first 8 digits of the ODFI's routing number followed by
	// 7 digits chosen by the ODFI or software vendor.
	//
	// Sequentual or random numbers can be chosen. The only requirement of Nacha is unique trace
	// numbers within a batch and file.
	//
	// Trace Numbers are included in each Entry Detail Record, Corporate Entry Detail Record,
	// and addenda Record.
	//
	// In association with the Batch Number, transmission (File Creation) Date,
	// and File ID Modifier, the Trace Number uniquely identifies an entry within a given file.
	//
	// For addenda Records, the Trace Number will be identical to the Trace Number
	// in the associated Entry Detail Record, since the Trace Number is associated
	// with an entry or item rather than a physical record.
	//
	// Use TraceNumberField for a properly formatted string representation.
	TraceNumber string `json:"traceNumber,omitempty"`
	// Addenda02 for use with StandardEntryClassCode MTE, POS, and SHR

	// Addenda05 for use with StandardEntryClassCode: ACK, ATX, CCD, CIE, CTX, DNE, ENR, WEB, PPD, TRX.

	// Addenda98 for user with Notification of Change

	// Addenda98 for user with Refused Notification of Change

	// Addenda99 for use with Returns

	// Addenda99Contested for use with Contested Dishonored Returns

	// Addenda99Dishonored for use with Dishonored Returns

	// Category defines if the entry is a Forward, Return, or NOC

	// validator is composed for data validation

	// converters is composed for ACH to golang Converters
	moov_io_ach_converters
}

type moov_io_ach_converters struct{}

File: translations_only/fail/go/ach/moov_io_ach_EntryDetail_TraceNumberField/gemini/Restart/moov_io_ach_EntryDetail_TraceNumberField.rs

use std::collections::HashMap;
use std::fmt::Write;
use std::str::FromStr;

#[derive(Debug)]
pub struct MoovIoAchEntryDetail {
    pub trace_number: String,
    pub addenda_02: Option<String>,
    pub addenda_05: Option<String>,
    pub addenda_98: Option<String>,
    pub addenda_98_for_user_with_refused_notification_of_change: Option<String>,
    pub addenda_99: Option<String>,
    pub addenda_99_contested: Option<String>,
    pub addenda_99_dishonored: Option<String>,
    pub category: String,
}

impl MoovIoAchEntryDetail {
    pub fn trace_number_field(&self) -> String {
        self.string_field(&self.trace_number, 15)
    }

    fn string_field(&self, s: &str, max: usize) -> String {
        let ln = s.chars().count();
        if ln > max {
            s[..max].to_string()
        } else {
            let mut pad = String::with_capacity(max - ln);
            for _ in 0..(max - ln) {
                pad.push('0');
            }
            pad + s
        }
    }
}

pub fn moov_io_ach_populate_map(max: usize, zero: &str) -> HashMap<usize, String> {
    let mut out = HashMap::with_capacity(max);
    for i in 0..max {
        out.insert(i, zero.repeat(i));
    }
    out
}

#[derive(Debug)]
pub struct MoovIoAchConverters {}

impl MoovIoAchConverters {
    pub fn string_field(&self, s: &str, max: usize) -> String {
        let ln = s.chars().count();
        if ln > max {
            s[..max].to_string()
        } else {
            let mut pad = String::with_capacity(max - ln);
            for _ in 0..(max - ln) {
                pad.push('0');
            }
            pad + s
        }
    }
}

pub fn moov_io_ach_entry_detail_from_str(s: &str) -> Result<MoovIoAchEntryDetail, String> {
    let mut entry_detail = MoovIoAchEntryDetail {
        trace_number: String::new(),
        addenda_02: None,
        addenda_05: None,
        addenda_98: None,
        addenda_98_for_user_with_refused_notification_of_change: None,
        addenda_99: None,
        addenda_99_contested: None,
        addenda_99_dishonored: None,
        category: String::new(),
    };

    let mut lines = s.lines();
    while let Some(line) = lines.next() {
        if line.starts_with("TraceNumber") {
            entry_detail.trace_number = line[11..].trim().to_string();
        } else if line.starts_with("Addenda02") {
            entry_detail.addenda_02 = Some(line[9..].trim().to_string());
        } else if line.starts_with("Addenda05") {
            entry_detail.addenda_05 = Some(line[9..].trim().to_string());
        } else if line.starts_with("Addenda98") {
            entry_detail.addenda_98 = Some(line[9..].trim().to_string());
        } else if line.starts_with("Addenda98") {
            entry_detail.addenda_98_for_user_with_refused_notification_of_change =
                Some(line[9..].trim().to_string());
        } else if line.starts_with("Addenda99") {
            entry_detail.addenda_99 = Some(line[9..].trim().to_string());
        } else if line.starts_with("Addenda99") {
            entry_detail.addenda_99_contested = Some(line[9..].trim().to_string());
        } else if line.starts_with("Addenda99") {
            entry_detail.addenda_99_dishonored = Some(line[9..].trim().to_string());
        } else if line.starts_with("Category") {
            entry_detail.category = line[8..].trim().to_string();
        }
    }

    Ok(entry_detail)
}
